### ISSUE TITLE
Interoperability between RTP over QUIC streams and plain RTP

### ISSUE THREAD

**Issue by murillo128** [mengelbart/rtp-over-quic-draft#42]
If I understood the draft correctly, basically when sending a frame with RTP over QUIC using streams, it will be "packetized" considering the MTU being infinite, that is, the whole frame would be sent as payload of the RTP packet (while still adhering to RTP packetization rules, I assume).

This would cause issues when trying to do a gateway that bridges between "normal" RTP and the QUIC stream version, as it will have to packetize/depacketize the whole frame again.

One potential solution is to still generate multiple RTP packets per frame and send them using RFC4571 over a single QUIC stream:

```
|len 1|RTP Packet1|len 2|RTP Packet2|...|len n|RTP PacketN|
```

This would solve the interoperability issues, although will have a higher bandwdith overhead.

**Comment by aboba** [comment]
I believe there are two sets of interop issues: 

1. - Interop between RTP over QUIC datagrams and RTP over UDP
2. - Interop between RTP over QUIC frame/stream transport and RTP over UDP

MTU issues can arise in either case. For example, RTP over QUIC datagrams may have a smaller MTU than RTP over UDP so RTP/UDP packets might need to be repacketized into multiple RTP/QUIC datagrams. 

To enable a translator to operate without repacketization (which could require codec-specific knowledge), the RTP/UDP MTU needs to be set smaller, so as to allow the translated RTP/QUIC datagrams to stay within the MTU. 

Similarly, RTP/QUIC stream transport could be configured with a non-infinite MTU, so as to allow a translator to translate to RTP/UDP without repacketizing.

**Comment by murillo128** [comment]
What is QUIC datagram overhead? i doubt #1 will ever be an issue in real life, while i doubt that QUIC over stream will be configured for non-infinite MTU as it defeat the whole pourpose of using streams.

**Comment by aboba** [comment]
The RTP Topologies section is there to describe topologies, issues and potential solutions.  The topologies include use cases such as conferencing where the SFU and browser clients could be designed to work together. 

In a use case where the conferencing server wants to support both RTP over QUIC and RTP/UDP clients, how can this be accomplished? In such a use case, the topology section needs to think about not only what a client should send, but also what it would need to be able to receive. 

If the conferencing server is an SFU and is willing to parse the payload, it can re-packetize from RTP over QUIC streams to RTP/UDP.  But it may not want to do this.  For example, if the SFU wants to support multiple codecs (e.g. VP8, VP9, H.264, AV1) it might want to handle forwarding in a generic way rather than include code to parse each codec. 

To be able to enable the SFU to do generic forwarding, the client can either configure a non-infinite MTU for RTP over QUIC streams or can send RTP over QUIC datagrams. Clients that only support RTP/UDP (e.g. browsers that only support WebRTC) could be handled by translating to RTP over QUIC streams if the alternative (RTP over QUIC datagrams) would result in fragmentation.

**Comment by mengelbart** [comment]
I opened #43 to add a length field and allow multiple RTP packets per stream as you suggested, @murillo128.

One feature I removed by doing this is the ability of a receiver to request to cancel a stream by sending a `STOP_SENDING` frame, if the receiver knows that a packet is no longer needed. This doesn't work, if there are other packets following on the same stream and the receiver does not know if those are still required. We could add it back if we still require senders to open new streams for each ADU, but since they might be fragmented into multiple RTP packets, I am not sure if it is worth keeping this restriction only to enable receivers to cancel streams.

What do you think about this?


@aboba I agree that RTP over QUIC Datagrams and RTP over UDP may still lead to interop issues, but I don't know if there is anything we can do about it except to document it in the topology section?

**Comment by mengelbart** [comment]
I merged #43, which adds the length field and allows multiple packets per stream, and added an issue for the `STOP_SENDING` frame problem (#45). I also documented the second MTU interop issue in the new topology section. Closing this issue for now, but feel free to reopen or create new issues if I missed something.

**Comment by aboba** [comment]
I think that the important thing is to be clear about what topologies are tractable and which are more difficult.  Conferencing servers are quite different from generic translators; the former often have the codec-specific knowledge to re-packetize, whereas translators do not.
